### PR TITLE
[ci] Add Jenkinsfile

### PR DESCRIPTION
**Pull request by woshilapin**
This is only a try to branch Jenkins as our new CI as Github Actions is showing some limitations. Note that this is only premises for the complete work, I didn't check the new configuration for the CI was exhaustively doing the same operations that the current CI.

### CODE REVIEW

**Review by pascal-vs — APPROVED** [review]

**Comment by woshilapin** [comment]
I've added a `makefile` to the project. This `makefile` tries to be a little smart so local developers can use it without Docker. Basically, this is how it works.

By default, any `make` command will run inside a Docker container. For a new developer, this can come handy since the new developer would only need Docker to be installed to start working. The problem with the Docker solution is that a build takes longer than if you have `rustup` and `cargo` installed on your machine (because containers won't cache anything: downloading of crates, compilation of some crates if you're using `sccache`, etc.).

To be able to work with your local tools, the `makefile` adapts from the `LOCAL_DEV` variable environment. If you declare a local environment variable `LOCAL_DEV=true`, all the command will run local (not inside a Docker container). Putting `export LOCAL_DEV=true` in your `.bashrc` should work but you might not want to create this variable for all terminals in every situation. So the solution is to create a local file `.env` that contains `LOCAL_DEV=true`. The `makefile` is reading that file.

Note that if the `.env` file doesn't exist, `make` won't fail because I'm using `-include` instead of `include` (see [documentation](https://www.gnu.org/software/make/manual/make.html#Include)).

**Comment by woshilapin** [comment]
I added the `help` target to autodocument the Makefile.

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+		stage ('Quality checks') {
+			parallel 'format': {
+				sh 'make format'
+			}, 'clippy': {
+				sh 'make lint'
+			}
+		}
+        stage ('Tests') {
+            sh 'make test'
+        }
```
Comment (ArnaudOggy) [inline Jenkinsfile]:
```suggestion
        stage ('Quality checks') {
	    parallel 'format': {
	        sh 'make format'
	    }, 'clippy': {
	        sh 'make lint'
	    }
        }
        stage ('Tests') {
            sh 'make test'
        }
```
indentation

File: makefile [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+format:
+	${SH} "rustup component add rustfmt && cargo fmt --all -- --check"
+
+clippy: lint
+
+lint:
+	${SH} "rustup component add clippy && cargo clippy --workspace --all-features -- --warn clippy::cargo --allow clippy::multiple_crate_versions --deny warnings"
+
+test:
+	${SH} "cargo test"
```
Comment (ArnaudOggy) [inline makefile]: harmonize with clippy (features, workspaces, ...)

File: makefile [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# By default, this makefile will run everything inside a Docker container.
+# If you're a developer and want to run everything with your local `cargo`:
+# - Creates a `.env` file in the same folder as this `makefile`
+# - Add a line in `.env` with `LOCAL_DEV=true`
+-include .env
+
+ifeq ($(LOCAL_DEV),true)
+	SH=sh -c
+else
+	SH=docker run --rm --user "$(id -u)":"$(id -g)" --volume "${PWD}":/tmp/skip_error --workdir /tmp/skip_error rust:latest sh -c
```
Comment (ArnaudOggy) [inline makefile]: put a generic name (instead of skip_error)

**Comment by ArnaudOggy** [comment]
@woshilapin you must have forgotten to push :smile:

File: makefile [hunk 4]
```diff
@@ -0,0 +1,24 @@
+# By default, this makefile will run everything inside a Docker container.
+# If you're a developer and want to run everything with your local `cargo`:
+# - Creates a `.env` file in the same folder as this `makefile`
+# - Add a line in `.env` with `LOCAL_DEV=true`
+-include .env
+
+ifeq ($(LOCAL_DEV),true)
+	SH=sh -c
```
Comment (pbougue) [inline makefile]: Here is a proposal to use the local shell : #6 

And I can't help but think that we should use the docker command only on CI (and skip the use as dev without a rust installed). So remove that whole `SH` (or function in my PR).
Same for the install of rustfmt.

Maybe script-helpers could be in makefile (`setup` and maybe `docker_setup` targets), but it's OK to expect the platform to be ready before running make targets IMO.

-> Detailed in a separate comment

File: Jenkinsfile (hunk 1 above)
Comment (pbougue) [reply Jenkinsfile]: Still some mix of tab and spaces

**Comment by woshilapin** [comment]
> @woshilapin you must have forgotten to push smile

I did indeed because I forgot this branch was not on my fork (I did on upstream because I was not sure Jenkins would pick it up). Should be fixed now.

File: Jenkinsfile (hunk 1 above)
Comment (woshilapin) [reply Jenkinsfile]: Yep, see below my "problem" of not pushing the update. Should be fixed now.

File: makefile (hunk 2 above)
Comment (woshilapin) [reply makefile]: Yep, I did that, just forgot to push. Should be ok now.

File: makefile (hunk 3 above)
Comment (woshilapin) [reply makefile]: I renamed it `workspace`.

**Comment by woshilapin** [comment]
I added a commit to silence the command calls which were displaying long and cryptic messages to the user (see 51f63ce16f66a28af51ae597752e67e403575c4f).

I also tried something additional, to display something better to the user. Tell me if you like it. It's a different commit so I can dump it if you don't care showing something to the user (see b5a2bac049b80730d947f902710deb72d70d3562).

**Review by pbougue — APPROVED** [review]
Not sharing the point-of-view on the role of this Makefile: 
* to me the need is just to centralize cmd to be run by CI and a contributor.
* I wouldn't complicate that much the makefile to ease the time-to-first-checks for contributor (16/28 loc for that): I would move docker part mostly to Jenkinsfile.

Although the work is great and it's way better with that IMO, so :heavy_check_mark: for that.

**Review by ArnaudOggy — APPROVED** [review]

**Pull request merged**